class.jetpack-settings-sync.php

```diff
@@ -0,0 +1,338 @@
+				if ( isset( self::$mock_options[ $name ][ 'callback_args' ] ) &&
+					 is_array( self::$mock_options[ $name ][ 'callback_args' ] ) ) {
+					$args = self::$mock_options[ $name ][ 'callback_args' ];
+				}
+
+				$data = call_user_func_array( self::$mock_options[ $name ][ 'callback' ], $args );
+
+				return self::validate( $data, self::$mock_options[ $name ][ 'type' ] );
+
+			} else {
+				new WP_Error( json_encode( self::$mock_options[ $name ][ 'callback' ] ) . ' can not be called' );
+				return null;
+			}
+
+		// Constants
+		} elseif( is_array( self::$constants ) && array_key_exists( $name, self::$constants) ) {
+
+			if ( defined( $name ) ) {
+				return self::validate( constant( $name ) , self::$constants[ $name ] );
+			} else {
+				return null;
+			}
+		}
+
+		return null;
+	}
+	/**
+	 * Gets all the settings or just some.
+	 *
+	 * @param  string or array $type could be 'all', 'options', 'mock_options', 'constants' it can also be an array of setting names
+	 * @return array $settings associative array of settings bases on type
+	 */
+	static function get_all( $type = null ) {
+
+		$all_settings = array();
+		$data = array();
+
+		if ( is_array( $type ) ) {
+			$all_settings = $type;
+		} else {
+			switch( $type ) {
+			case 'options':
+				$all_settings = array_keys( self::$options );
+				break;
+
+			case 'mock_options':
+				$all_settings = array_keys( self::$mock_options );
+				break;
+
+			case 'constants':
+				$all_settings = array_keys( self::$constants );
+				break;
+
+			case 'all':
+			default:
+				$all_settings = array_merge( array_keys( self::$options ), array_keys( self::$mock_options ), array_keys( self::$constants ) );
+				break;
+
+			}
+		}
+
+		foreach ( $all_settings  as $name  ) {
+				$data[ $name ] = self::get( $name );
+		}
+
+		return $data;
+	}
+
+	static function update( $name, $data ) {
+
+		if ( ! isset( self::$options[$name] ) )
+			return;
+
+		$data = self::validate( $data, self::$options[$name] );
+		return update_option( $name, $data );
+
+	}
+
+	static function validate( $data, $type ) {
+
+		if ( is_null( $data ) ) {
+			return $data;
+		}
+
+		switch( $type ) {
+			case 'bool':
+				return boolval( $data );
+
+			case 'url':
+				return esc_url( $data );
+
+			case 'on':
+				return ( $data == 'on' ? true : false );
+				break;
+
+			case 'closed':
+				return ( $data != 'closed' ? true : false );
+
+			case 'string':
+				return strval( $data );
+
+			case 'int':
+				return ( is_numeric( $data ) ? intval( $data ) : 0 );
+
+			case 'array':
+				return ( is_array( $data ) ? $data : array() );
+		}
+
+		if (  ! is_array( $type ) && 'regex:' == substr( $type, 0, 5 ) ) {
+
+			return ( preg_match( substr( $type, 6 ), $data ) ? $data : null );
+
+		} elseif ( is_array( $type ) && ! isset( $type[0]) ) {
+
+			foreach ( $type as $item => $check ) {
+				$data[ $item ] = self::validate( $data[ $item ], $check );
+			}
+			return $data;
+
+		} elseif ( is_array( $type ) && isset( $type[0] ) ) {
```

Comment:
I would put this inside the previous `elseif` to avoid the double `is_array( $type )` check